umbrella-feeds/src/lib.rs

```diff
@@ -0,0 +1,732 @@
+        );
+
+        *stored_price_data = new_price_data;
+    }
+
+    Ok(())
+}
+
+/// View function that returns the key hash/name of this contract.
+#[receive(
+    contract = "umbrella_feeds",
+    name = "getName",
+    return_value = "HashSha2256",
+    crypto_primitives
+)]
+fn get_name<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    _host: &impl HasHost<State<S>, StateApiType = S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> ReceiveResult<HashSha2256> {
+    let key_hash = crypto_primitives.hash_sha2_256(NAME.as_bytes()).0;
+
+    Ok(HashSha2256(key_hash))
+}
+
+/// View function that returns many price data. It throws if price feed does not exist.
+#[receive(
+    contract = "umbrella_feeds",
+    name = "getManyPriceData",
+    parameter = "Vec<HashSha2256>",
+    return_value = "Vec<PriceData>"
+)]
+fn get_many_price_data<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ReceiveResult<Vec<PriceData>> {
+    let key_hashes: Vec<HashSha2256> = ctx.parameter_cursor().get()?;
+
+    let mut price_data = Vec::with_capacity(key_hashes.len());
+
+    for key_hash in key_hashes {
+        price_data.push(
+            *host
+                .state()
+                .prices
+                .get(&key_hash)
+                .ok_or(CustomContractError::FeedNotExist)?,
+        );
+    }
+
+    Ok(price_data)
+}
+
+/// View function that returns many price data. This function does NOT throw if the price feed does not exists.
+/// The original solidity function returned a default PriceData instead of throwing an error when the price feed did not exsts in this contract.
+/// This behavior (also not typical for Rust) is mimicked.
+#[receive(
+    contract = "umbrella_feeds",
+    name = "getManyPriceDataRaw",
+    parameter = "Vec<HashSha2256>",
+    return_value = "Vec<PriceData>"
+)]
+fn get_many_price_data_raw<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ReceiveResult<Vec<PriceData>> {
+    let key_hashes: Vec<HashSha2256> = ctx.parameter_cursor().get()?;
+
+    let mut price_data = Vec::with_capacity(key_hashes.len());
+
+    for key_hash in key_hashes {
+        price_data.push(
+            host.state()
+                .prices
+                .get(&key_hash)
+                .map(|s| *s)
+                .unwrap_or_else(PriceData::default),
+        );
+    }
+
+    Ok(price_data)
+}
+
+/// View function that returns the price data of one price feed. This function does NOT throw if the price feed does not exists.
+/// The original solidity function returned a default PriceData instead of throwing an error when the price feed did not exsts in this contract.
+/// This behavior (also not typical for Rust) is mimicked.
+#[receive(
+    contract = "umbrella_feeds",
+    name = "prices",
+    parameter = "HashSha2256",
+    return_value = "PriceData"
+)]
+fn prices<S: HasStateApi>(
```

Comment:
Can you please explain why this function is added?

As I commented elsewhere, the default pricedata seems arbitrary. We really need to have a strong reason for including such an error-prone function.

Reply:
As mentioned, umbrella is a multi-chain protocol that incorporates EVM and non-EVM chains. This function originated from solidity and I agree with you that this is not a recommended pattern (to return default values) in Rust, we also need to understand umbrella's position of staying compatible across chains as much as possible. I have informed them about it in the comments and added it here in the PR again in the link below. It would be best if they make decisions about which of the getter functions they want to keep, we have provided them all the getter functions that originally existed.

https://github.com/Concordium/concordium-umbrella-oracle/pull/1#issuecomment-1738819711